File: app/src/main/java/com/github/goldy1992/mp3player/client/callbacks/subscription/MediaIdSubscriptionCallback.java
package com.github.goldy1992.mp3player.client.callbacks.subscription;

import android.os.Handler;
import android.support.v4.media.MediaBrowserCompat;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import com.github.goldy1992.mp3player.client.MediaBrowserResponseListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Named;

public class MediaIdSubscriptionCallback extends MediaBrowserCompat.SubscriptionCallback {

    private static final String LOG_TAG = "SUBSCRIPTION_CALLBACK";
    private Handler handler;
    private final Map<String, Set<MediaBrowserResponseListener>> mediaBrowserResponseListeners;

    @Inject
    public MediaIdSubscriptionCallback(@Named("worker") Handler handler) {
        this.mediaBrowserResponseListeners = new HashMap<>();
        this.handler = handler;

    }

    @Override
    public void onChildrenLoaded(@NonNull String parentId, @NonNull List<MediaBrowserCompat.MediaItem> children) {
        handler.post( () -> {
            ArrayList<MediaBrowserCompat.MediaItem> childrenArrayList = new ArrayList<>(children);
            Set<MediaBrowserResponseListener> listenersToNotify = mediaBrowserResponseListeners.get(parentId);
            if (null != listenersToNotify) {
                for (MediaBrowserResponseListener listener : listenersToNotify) {
                    listener.onChildrenLoaded(parentId, childrenArrayList);
                }
            }
        });
    }

    public synchronized void registerMediaBrowserResponseListener(String key , MediaBrowserResponseListener listener) {
        if (mediaBrowserResponseListeners.get(key) == null) {
            mediaBrowserResponseListeners.put(key, new HashSet<>());
        }
        mediaBrowserResponseListeners.get(key).add(listener);
    }

    @VisibleForTesting
    public Map<String, Set<MediaBrowserResponseListener>> getMediaBrowserResponseListeners() {
        return mediaBrowserResponseListeners;
    }
}

File: app/src/main/java/com/github/goldy1992/mp3player/client/callbacks/subscription/MediaIdSubscriptionCallback.kt
package com.github.goldy1992.mp3player.client.callbacks.subscription

import android.os.Handler
import android.support.v4.media.MediaBrowserCompat
import androidx.annotation.VisibleForTesting
import com.github.goldy1992.mp3player.client.MediaBrowserResponseListener
import java.util.*
import javax.inject.Inject
import javax.inject.Named

class MediaIdSubscriptionCallback @Inject constructor(@Named("worker") handler: Handler) : MediaBrowserCompat.SubscriptionCallback() {
    private val handler: Handler
    private val mediaBrowserResponseListeners: MutableMap<String, MutableSet<MediaBrowserResponseListener>>
    override fun onChildrenLoaded(parentId: String, children: List<MediaBrowserCompat.MediaItem>) {
        handler.post {
            val childrenArrayList = ArrayList(children)
            val listenersToNotify: Set<MediaBrowserResponseListener>? = mediaBrowserResponseListeners[parentId]
            if (null != listenersToNotify) {
                for (listener in listenersToNotify) {
                    listener.onChildrenLoaded(parentId, childrenArrayList)
                }
            }
        }
    }

    @Synchronized
    fun registerMediaBrowserResponseListener(key: String, listener: MediaBrowserResponseListener?) {
        if (mediaBrowserResponseListeners[key] == null) {
            mediaBrowserResponseListeners[key] = HashSet()
        }
        mediaBrowserResponseListeners[key]!!.add(listener!!)
    }

    @VisibleForTesting
    fun getMediaBrowserResponseListeners(): Map<String, MutableSet<MediaBrowserResponseListener>> {
        return mediaBrowserResponseListeners
    }

    companion object {
        private const val LOG_TAG = "SUBSCRIPTION_CALLBACK"
    }

    init {
        mediaBrowserResponseListeners = HashMap()
        this.handler = handler
    }
}